server/src/com/thoughtworks/go/config/CachedGoPartials.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright 2015 ThoughtWorks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.thoughtworks.go.config;
+
+import com.thoughtworks.go.config.remote.PartialConfig;
+import com.thoughtworks.go.config.remote.RepoConfigOrigin;
+import org.joda.time.DateTime;
+import org.springframework.stereotype.Component;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static com.thoughtworks.go.util.ExceptionUtils.bomb;
+
+@Component
+public class CachedGoPartials {
+    public List<PartialConfig> lastValidPartials() {
+        return getPartialConfigs(fingerprintToLatestValidConfigMap);
+    }
+
+    public List<PartialConfig> lastKnownPartials() {
+        return getPartialConfigs(fingerprintToLatestKnownConfigMap);
+    }
+
+    private List<PartialConfig> getPartialConfigs(Map<String, UpdatedPartial> map) {
+        List<PartialConfig> list = new ArrayList<>();
+        for (UpdatedPartial partialConfig : map.values()) {
+            list.add(partialConfig.partialConfig);
+        }
+        return list;
+    }
+
+    public void removeKnown(String fingerprint) {
+        if (fingerprintToLatestKnownConfigMap.containsKey(fingerprint)) {
+            fingerprintToLatestKnownConfigMap.remove(fingerprint);
+        }
+    }
+
+    public PartialConfig getValid(String fingerprint) {
+        return getPartialConfig(fingerprint, fingerprintToLatestValidConfigMap);
+    }
+
+    public PartialConfig getKnown(String fingerprint) {
+        return getPartialConfig(fingerprint, fingerprintToLatestKnownConfigMap);
+    }
+
+    private PartialConfig getPartialConfig(String fingerprint, Map<String, UpdatedPartial> map) {
+        if (map.containsKey(fingerprint)) {
+            return map.get(fingerprint).partialConfig;
+        }
+        return null;
+    }
+
+    public void removeValid(String fingerprint) {
+        if (fingerprintToLatestValidConfigMap.containsKey(fingerprint)) {
+            fingerprintToLatestValidConfigMap.remove(fingerprint);
+        }
+    }
+
+
+    public void addOrUpdate(String fingerprint, PartialConfig newPart) {
+        fingerprintToLatestKnownConfigMap.put(fingerprint, new UpdatedPartial(newPart, new DateTime()));
+    }
+
+    public void markAsValid(String fingerprint, PartialConfig newPart) {
+        DateTime lastUpdated = fingerprintToLatestKnownConfigMap.get(fingerprint).lastUpdated;
```

Comment:
Considering that this is currently throwing NullPointerException maybe it should throw some nicer message when `markAsValid` is called with a fingerprint that does not exist in map